### PR TITLE
Make AddControllersAsServices not overwrite existing IControllerTypeProvider

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #4086 

@rynowak / @Eilon

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllersAsServices.cs [hunk 1]
```diff
@@ -26,7 +26,14 @@ public static void AddControllersAsServices(IServiceCollection services, IEnumer
...
 
-            var controllerTypeProvider = new StaticControllerTypeProvider();
+            StaticControllerTypeProvider controllerTypeProvider = null;
+
+            var existingServiceDescriptor = services.SingleOrDefault(s => s.ServiceType == typeof(IControllerTypeProvider));
+            controllerTypeProvider = existingServiceDescriptor?.ImplementationInstance as StaticControllerTypeProvider;
+            if (controllerTypeProvider == null)
+            {
+                controllerTypeProvider = new StaticControllerTypeProvider();
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllersAsServices.cs]: You could simplify this a bit more with linq. 
```
services.Where(s => s.ServiceType...).Select(s => s.ImplementationType).SingleOrDefault() ?? new ...;
```

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/ControllerFromServicesTests.cs [hunk 2]
```diff
@@ -104,5 +104,18 @@ public ControllerFromServicesTest(MvcTestFixture<ControllersFromServicesWebSite.
...
+        public async Task AddControllersAsServices_MultipleCalls_DoesNotReplacePreviousProvider()
+        {
+            // Arrange
+            var expected = "1";
+
+            // Act
+            var response = await Client.GetStringAsync("http://localhost/another/");
+
+            // Assert
+            Assert.Equal(expected, response);
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/ControllerFromServicesTests.cs]: Why is this a functional test?

Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/ControllerFromServicesTests.cs]: Makes a little more sense, but I'd like to see a unit test all the same.

**Comment by rynowak** [comment]
:watch:

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/ControllerFromServicesTests.cs (hunk 2 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/ControllerFromServicesTests.cs]: I couldn't find any existing unit tests for `ControllerAsServices`.

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllersAsServices.cs [hunk 3]
```diff
@@ -26,7 +26,14 @@ public static void AddControllersAsServices(IServiceCollection services, IEnumer
                 throw new ArgumentNullException(nameof(types));
             }
 
-            var controllerTypeProvider = new StaticControllerTypeProvider();
+            StaticControllerTypeProvider controllerTypeProvider = null;
+
+            var existingServiceDescriptor = services.SingleOrDefault(s => s.ServiceType == typeof(IControllerTypeProvider));
```
Comment (pranavkm) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllersAsServices.cs]: `FirstOrDefault`? We don't actually care if there's more than one `IControllerTypeProvider`, do we?

Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllersAsServices.cs]: :+1: Actually `ServiceCollection.Replace()` uses `FirstOrDefault` internally.

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/ControllerFromServicesTests.cs (hunk 2 above)
Comment (javiercn) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/ControllerFromServicesTests.cs]: @ajaybhargavb Add a file and add the tests related to your feature :smile:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllersAsServices.cs (hunk 1 above)
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllersAsServices.cs]: Alternative,

```
if (!existingServiceDescriptor?.ImplementationInstance is StaticControllerTypeProvider)
{
    controllerTypeProvider = new StaticControllerTypeProvider();
}
```

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/ControllerFromServicesTests.cs (hunk 2 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/ControllerFromServicesTests.cs]: Added.

**Comment by ajaybhargavb** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllersAsServices.cs [hunk 4]
```diff
@@ -26,7 +26,12 @@ public static void AddControllersAsServices(IServiceCollection services, IEnumer
                 throw new ArgumentNullException(nameof(types));
             }
 
-            var controllerTypeProvider = new StaticControllerTypeProvider();
+            StaticControllerTypeProvider controllerTypeProvider = null;
+
+            controllerTypeProvider = services
+                .Where(s => s.ServiceType == typeof(IControllerTypeProvider))
+                .Select(s => s.ImplementationInstance as StaticControllerTypeProvider).FirstOrDefault()
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllersAsServices.cs]: your condition here is just `.OfType<Static...>(). Linq already has that :)

**Comment by rynowak** [comment]
:shipit:

**Pull request merged**